sql/src/main/java/io/deephaven/sql/TypeAdapter.java

```diff
@@ -0,0 +1,132 @@
+/**
+ * Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+ */
+package io.deephaven.sql;
+
+import io.deephaven.qst.column.header.ColumnHeader;
+import io.deephaven.qst.table.TableHeader;
+import io.deephaven.qst.type.ArrayType;
+import io.deephaven.qst.type.BooleanType;
+import io.deephaven.qst.type.ByteType;
+import io.deephaven.qst.type.CharType;
+import io.deephaven.qst.type.CustomType;
+import io.deephaven.qst.type.DoubleType;
+import io.deephaven.qst.type.FloatType;
+import io.deephaven.qst.type.GenericType;
+import io.deephaven.qst.type.InstantType;
+import io.deephaven.qst.type.IntType;
+import io.deephaven.qst.type.LongType;
+import io.deephaven.qst.type.PrimitiveType;
+import io.deephaven.qst.type.ShortType;
+import io.deephaven.qst.type.StringType;
+import io.deephaven.qst.type.Type;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeFactory;
+import org.apache.calcite.rel.type.RelDataTypeFactory.Builder;
+import org.apache.calcite.sql.type.SqlTypeName;
+
+import java.util.Objects;
+
+final class TypeAdapter implements Type.Visitor, GenericType.Visitor, PrimitiveType.Visitor {
+
+    public static RelDataType of(Type<?> type, RelDataTypeFactory typeFactory) {
+        final TypeAdapter adapter = new TypeAdapter(typeFactory);
+        type.walk(adapter);
+        return Objects.requireNonNull(adapter.out);
+    }
+
+    public static RelDataType of(TableHeader tableHeader, RelDataTypeFactory typeFactory) {
+        final Builder builder = new Builder(typeFactory);
+        for (ColumnHeader<?> columnHeader : tableHeader) {
+            builder.add(columnHeader.name(), of(columnHeader.componentType(), typeFactory));
+        }
+        return builder.build();
+    }
+
+    private final RelDataTypeFactory typeFactory;
+    private RelDataType out;
+
+    private TypeAdapter(RelDataTypeFactory typeFactory) {
+        this.typeFactory = Objects.requireNonNull(typeFactory);
+    }
+
+    @Override
+    public void visit(PrimitiveType<?> primitiveType) {
+        primitiveType.walk((PrimitiveType.Visitor) this);
+    }
+
+    @Override
+    public void visit(BooleanType booleanType) {
+        out = create(SqlTypeName.BOOLEAN);
+    }
+
+    @Override
+    public void visit(ByteType byteType) {
+        out = create(SqlTypeName.TINYINT);
```

Comment:
I think this is an incorrect mapping. We have a symmetry problem, no matter what, without richer types.